### PR TITLE
Clarify documentation guideline - do no include empty xml documents

### LINKED ISSUE #10857: Clarify documentation guideline - do no include empty xml docs

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/10857

Regression? Last working version:

## Description

Do not include empty xml documents as they provide no value and just take up space.
Note that empty docs *are not* shown by intellisense.

## PR Checklist

- [ ] PR has a meaningful title
- [ ] PR has a linked issue.
- [ ] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [ ] N/A

### CODE REVIEW

**Review by donnie-msft — CHANGES_REQUESTED** [review]
Oh a PR already. Blocking on my question on Issue 😋
https://github.com/NuGet/Home/issues/10857#issuecomment-839170905

File: docs/coding-guidelines.md
Comment (donnie-msft) [inline docs/coding-guidelines.md:347]: How about an argument? `string fruitName`

Then the comment block examples of:
-  I don't want to mention any specifics (ie, what's the minimum acceptable text beyond empty string?)
- I want to mention only pre/post conditions
- I want to write a very detailed description.

**Comment by nkolev92** [comment]
@donnie-msft let's just have the discussion here. We can tweak the guidance to whatever the outcome is.

File: docs/coding-guidelines.md
Comment (nkolev92) [reply docs/coding-guidelines.md:347]: I updated it again, ptal.

I'm trying not to make this description too long. 
Any suggestions? 

I think the guidance is as simple as don't declare elements that are empty since the intellisense ignores them and they provide no tangible benefit.

Comment (donnie-msft) [reply docs/coding-guidelines.md:347]: Yes, I made suggestions on the Issue. It isn't clear if you want the element to be removed, or if there should be a default text.

> Can we make this more explicit? eg, Specify all parameters, at a minimum with their pre- & post-conditions. If nothing of note, should we write a basic descriptions like, "The packageId"?
> 
> By not listing a parameter for a method with a parameter, are we not breaking the convention?

Comment (donnie-msft) [reply docs/coding-guidelines.md:347]: I see in your update that you want the elements, but with a basic description. That's exactly what I asked :) sounds good

File: docs/coding-guidelines.md
Comment (donnie-msft) [inline docs/coding-guidelines.md]:
```suggestion
        /// <returns>A set of fruit types contained in this basket.</returns>
```

**Review by donnie-msft — DISMISSED** [review]

File: docs/coding-guidelines.md
Comment (nkolev92) [reply docs/coding-guidelines.md]: fixed.

File: docs/coding-guidelines.md
Comment (nkolev92) [reply docs/coding-guidelines.md:347]: Cool. I added an extra example to illustrate that omitting obvious params is ok.

File: docs/coding-guidelines.md
Comment (donnie-msft) [inline docs/coding-guidelines.md]: You added this after my review. I'm not sure about this. Now the XML comment doesn't match the signature :\

Comment (nkolev92) [reply docs/coding-guidelines.md]: The idea here is that you do all or none? 
You're basically assuming the names are all self descriptive enough?

Comment (donnie-msft) [reply docs/coding-guidelines.md]: Yes, all. No, it's okay if you want to require the names be repeated in the description. Just never seen arbitrarily missing arguments before.

Comment (zivkan) [reply docs/coding-guidelines.md]: I believe that the analyzers built into roslyn even warn when you have xmldoc that is missing for a parameter.

**Comment by zivkan** [comment]
Basically, this PR's suggestion is re-implementing Roslyn's analyzers in a coding style document.

Unfortunately, the relevant code analyzers have been disabled on some (most?) projects at the project level. I feel it's important to keep the analyzers & good API docs for the public APIs in the NuGet SDK projects since people outside of our own team may be calling those APIs and need to know how to use it correctly. I don't know how detailed xmldoc we want for our client projects.

Would it be worthwhile to stop blocking the analyzers instead of, or in addition to, this PR?

File: docs/coding-guidelines.md
Comment (nkolev92) [reply docs/coding-guidelines.md]: > Yes, all. No, it's okay if you want to require the names be repeated in the description. Just never seen arbitrarily missing arguments before.

@donnie-msft My suggestion is all or nothing. 
You either describe *all* parameters, or none. Same applies to *every* xml doc. 

I can add a better example to clarify it.

Comment (nkolev92) [reply docs/coding-guidelines.md]: @zivkan When I looked yesterday, I concluded that those analyzers were not installed and I didn't want to propose an addition of new analyzers.
There might be *a lot* of things to be suppressed if we were to do that.

Comment (zivkan) [reply docs/coding-guidelines.md]: Visual Studio provides a 3 click way to create suppressions for every instance in the project (maybe even solution?). It's not a high effort task, if that is the concern.

Comment (nkolev92) [reply docs/coding-guidelines.md]: I'm not worried about the amount of effort. 
I just don't want our suppression file to be half warnings for comments. 
There's more important things to tackle imo.

**Comment by nkolev92** [comment]
@donnie-msft I dismissed your review as it was on an earlier version of the change.

File: docs/coding-guidelines.md
Comment (donnie-msft) [inline docs/coding-guidelines.md]: Why do we want to encourage "missing params" or say it's correct? I believe other comments from zivkan and myself point toward this as bad practice.

Comment (nkolev92) [reply docs/coding-guidelines.md]: I'm suggesting *all params* or *no params*. 

It's the closest to how our current APIs are documented.

Comment (donnie-msft) [reply docs/coding-guidelines.md]: For new guidance, I'd prefer we always include all params, even if just a boilerplate string is used in the content.
However, it's just my preference and I won't  block on it.

Comment (nkolev92) [reply docs/coding-guidelines.md]: > However, it's just my preference and I won't block on it.

Let's get a 3rd opinion. 

@zivkan or someone else from @NuGet/nuget-client 

I'm happy to go either direction.

Comment (zkat) [reply docs/coding-guidelines.md]: If we can enforce it at a tooling level, I agree with Donnie that all **public** methods should have **all** params documented, plus the `return` if non-`void`.

I don't think we should waste time in code reviews over this, though.

**Review by zkat — APPROVED** [review]

File: docs/coding-guidelines.md
Comment (zivkan) [reply docs/coding-guidelines.md]: Personally, I don't like the global supressions file, because when working on the `cs` file, you have no idea that a warning was suppressed. This makes it more difficult to apply the "boy scout rule" of improving code close to what you're editing already. But this is off topic for this PR.

Having said that, I don't understand what harm having a global supressions file which is mostly ignoring XMLDoc 
issues causes. There is tooling that can help us apply this rule, meaning we don't have to worry about it during code reviews, as per [this comment](https://github.com/NuGet/NuGet.Client/pull/4051#discussion_r635643302). But if we keep the analyzer disabled, then the tooling won't even run, and then the only way this guideline can be enforced is if we do check for it manually when reviewing other people's PRs.  This guidelines document is fairly long already, and it's difficult to keep all the different guidelines in mind when writing code.

>  I concluded that those analyzers were not installed

I believe the analyzers are built into Roslyn, not additional packages that need to be referenced. For example [cs1591](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs1591) [cs1573](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs1573), [cs1574](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs1574), appear to be part of the compiler. The docs certainly don't suggest that they're part of a different package/analyzer.  However, in at least one project, we explicitly `NoWarn` them:

https://github.com/NuGet/NuGet.Client/blob/9500c09af1cb9a79dd9df43775922e467d7ebe2c/src/NuGet.Core/NuGet.Frameworks/NuGet.Frameworks.csproj#L10

Therefore any developer who unintentionally violates the warning will never know.

Comment (nkolev92) [reply docs/coding-guidelines.md]: > I believe the analyzers are built into Roslyn, not additional packages that need to be referenced. For example cs1591 cs1573, cs1574, appear to be part of the compiler. T

That explains why I could find them in the analyzers warnings :D 

22 projects suppress `cs1591`, so pretty much all of the ones that generate documentation files. 

I played with enabling some of the warnings (I really removed all suppressed warnings in NuGet.Common) and the analyzers enforce the following. 

* The existence of a summary element on public APIs.
* The existence of all or none param elements on public APIs.

What these analyzers do not enforce:
* Actual content in the summary or param elements.

So as is, those rules are *not as strict* as our guidance is, thus the guidance is actually necessary either way.

@zivkan are you suggesting removing this warning from the project files and suppress it in global suppressions instead?

Comment (zivkan) [reply docs/coding-guidelines.md]: Considering the main feedback we get about the NuGet SDK, the few times we get feedback about the NuGet SDK, is that there's a lack of documentation, I think removing the NoWarns will encourage developers creating public APIs to add docs. It might also encourage developers creating new APIs to think a little more about good API design, which I'd also consider a good thing. If it happens, that'd be a great unintended consequence, in my opinion.

I'm fine any way for the non-SDK projects. If we don't want public API XMLDoc for APIs that customers can never call (in our VS assemblies, for example), I won't complain.

However, since the analyzers don't enforce non-empty XMLDoc, then it will require a team culture of writing meaningful docs if we want to improve the NuGet SDK feedback. Therefore, we'd probably need to actually comment on it in code reviews, contrary to my interpretation of @zkat's other comment.

I'd also prefer inline, rather than global, suppressions. But I know you in particular don't like commits that make small modifications to a large number of files.

Comment (nkolev92) [reply docs/coding-guidelines.md]: CS* warnings cannot be suppressed in a global suppressions file.
Only CA ones can be suppressed: 
https://docs.microsoft.com/en-us/visualstudio/code-quality/in-source-suppression-overview?view=vs-2019#suppress-violations-using-a-global-suppression-file

Comment (nkolev92) [reply docs/coding-guidelines.md]: > I'm fine any way for the non-SDK projects. If we don't want public API XMLDoc for APIs that customers can never call (in our VS assemblies, for example), I won't complain.

I agree. No need to overdo it from the onset. 

> However, since the analyzers don't enforce non-empty XMLDoc, then it will require a team culture of writing meaningful docs if we want to improve the NuGet SDK feedback. Therefore, we'd probably need to actually comment on it in code reviews, contrary to my interpretation of @zkat's other comment.

I agree with @zivkan. Part of code reviews is actually reading the documentation as well. Docs existent is not our bar, it's accurate docs. 

> I'd also prefer inline, rather than global, suppressions. But I know you in particular don't like commits that make small modifications to a large number of files.

Given that there are ~22 files I'd rather that be handled in a separate work item. 
It'd be great if the API were documented instead of just suppressing them. 
We could also potentially remove some redundant APIs that we discover in that effort. 6.0/ Dev17 is as good as any for that type of change.

Comment (zivkan) [reply docs/coding-guidelines.md]: > It'd be great if the API were documented instead of just suppressing them.

I agree. I just didn't want the burden of documenting the APIs on the person who stops using NoWarn

**Review by zivkan — APPROVED** [review]

**Pull request merged**
